### PR TITLE
Add a CSR batched matrix format, CUDA, HIP and DPCPP kernels

### PR DESCRIPTION
**Pull request by pratikvn**
This PR adds a batched Csr matrix format, that stores the same sparsity pattern for all entries, but different values for each batch. Only simple and advanced apply to `batch::MultiVector` are supported for now.

### TODO

+ [x] Add CUDA/HIP kernels
+ [x] Add DPCPP kernels
+ [x] Add to batch::Bicgstab dispatch
+ [x] Update docs

### CODE REVIEW

File: common/cuda_hip/matrix/batch_csr_kernels.hpp.inc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+__device__ __forceinline__ void advanced_apply(
+    const ValueType alpha,
+    const gko::batch::matrix::csr::batch_item<const ValueType, IndexType>& mat,
+    const ValueType* const __restrict__ b, const ValueType beta,
+    ValueType* const __restrict__ x)
+{
+    const auto num_rows = mat.num_rows;
+    const auto val = mat.values;
+    const auto col = mat.col_idxs;
+    for (int tidx = threadIdx.x; tidx < num_rows; tidx += blockDim.x) {
```
Comment (upsj) [inline common/cuda_hip/matrix/batch_csr_kernels.hpp.inc:93]: nit: naming
```suggestion
    for (int row = threadIdx.x; row < num_rows; row += blockDim.x) {
```

File: common/cuda_hip/matrix/batch_csr_kernels.hpp.inc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+template <typename ValueType, typename IndexType>
+__device__ __forceinline__ void simple_apply(
+    const gko::batch::matrix::csr::batch_item<const ValueType, IndexType>& mat,
+    const ValueType* const __restrict__ b, ValueType* const __restrict__ x)
+{
+    const auto num_rows = mat.num_rows;
+    const auto val = mat.values;
+    const auto col = mat.col_idxs;
+    for (int tidx = threadIdx.x; tidx < num_rows; tidx += blockDim.x) {
```
Comment (upsj) [inline common/cuda_hip/matrix/batch_csr_kernels.hpp.inc:42]:
```suggestion
    for (int row = threadIdx.x; row < num_rows; row += blockDim.x) {
```

File: core/matrix/csr.cpp [hunk 3]
```diff
@@ -471,8 +471,9 @@ void Csr<ValueType, IndexType>::read(device_mat_data&& data)
     auto arrays = data.empty_out();
     this->row_ptrs_.resize_and_reset(size[0] + 1);
     this->set_size(size);
-    this->values_ = std::move(arrays.values);
-    this->col_idxs_ = std::move(arrays.col_idxs);
+    // Need to copy instead of move, as move into view data is not supported.
+    this->values_ = arrays.values;
+    this->col_idxs_ = arrays.col_idxs;
```
Comment (upsj) [inline core/matrix/csr.cpp:476]: This is a pessimization for many use cases, and defeats the purposes of this being an rvalue overload. It sounds like calling `copy_from` might be more appropriate here.

File: core/test/matrix/batch_csr.cpp [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+}
+
+
+TYPED_TEST(Csr, SparseMtxKnowsItsSizeAndValues)
+{
+    this->assert_equal_to_original_sparse_mtx(this->sp_mtx.get());
+}
+
+
+// TYPED_TEST(Csr, CanBeEmpty)
```
Comment (upsj) [inline core/test/matrix/batch_csr.cpp:164]: I don't think we should commit commented out code. you can disable the tests by prefixing them with `DISABLED_` IIRC

File: dpcpp/matrix/batch_csr_kernels.hpp.inc [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        temp = ::gko::kernels::dpcpp::reduce(
+            sg, temp, [](ValueType a, ValueType b) { return a + b; });
+
+
+        if (sg.get_local_id() == 0) {
+            c[row] = alpha * temp + beta * c[row];
+        }
+    }
```
Comment (upsj) [inline dpcpp/matrix/batch_csr_kernels.hpp.inc:83]: This is a different level of optimization than the other kernels. Is that intended?

File: omp/matrix/batch_ell_kernels.cpp [hunk 6]
```diff
@@ -37,7 +37,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 #include <ginkgo/core/base/batch_multi_vector.hpp>
-#include <ginkgo/core/matrix/batch_dense.hpp>
+#include <ginkgo/core/matrix/batch_ell.hpp>
```
Comment (upsj) [inline omp/matrix/batch_ell_kernels.cpp:40]: leftover?

**Review by upsj — DISMISSED** [review]
This PR is pretty late, so I only went over the implementation, not the tests. Most of it LGTM, there is just a behavioral change in here that I believe overshoots in fixing a specific edge case.

**Comment by pratikvn** [comment]
@upsj , this was not meant to be in the release. I had a bit of time, so decided to work on it. Thanks for reviewing, but this is still a WIP, that is why I havent requested any reviews or marked it ready-for-review  yet.

**Comment by upsj** [comment]
@pratikvn phew, thank you, that would have been a night shift otherwise 😄

File: core/matrix/csr.cpp (hunk 3 above)
Comment (pratikvn) [reply core/matrix/csr.cpp:476]: Yes, I agree. I was also planning on checking if the arrays are `is_owning()` and move if so, and copy otherwise

**Comment by pratikvn** [comment]
format!

**Comment by pratikvn** [comment]
format!

File: core/test/matrix/batch_csr.cpp [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+    auto mat1 =
+        gko::initialize<CsrMtx>({{1.0, 0.0, 0.0}, {0.0, 2.0, 0.0}}, this->exec);
+    auto bat_m = gko::batch::create_from_item<BatchCsrMtx>(
+        this->exec, std::vector<CsrMtx*>{mat1.get(), mat1.get(), mat1.get()},
+        2);
+
+    auto m =
+        gko::batch::create_from_item<BatchCsrMtx>(this->exec, 3, mat1.get(), 2);
+
+    GKO_ASSERT_BATCH_MTX_NEAR(bat_m.get(), m.get(), 1e-14);
```
Comment (MarcelKoch) [inline core/test/matrix/batch_csr.cpp:332]: Why does this need `eps > 0`? I think all operations are just copies.

File: core/test/matrix/batch_csr.cpp [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+        this->exec, std::vector<CsrMtx*>{mat1.get(), mat2.get()}, 2);
+    auto m_ref = gko::batch::create_from_item<BatchCsrMtx>(
+        this->exec,
+        std::vector<CsrMtx*>{mat1.get(), mat2.get(), mat1.get(), mat2.get(),
+                             mat1.get(), mat2.get()},
+        2);
+
+    auto m2 = gko::batch::duplicate<BatchCsrMtx>(this->exec, 3, m.get(), 2);
+
+    GKO_ASSERT_BATCH_MTX_NEAR(m2.get(), m_ref.get(), 1e-14);
```
Comment (MarcelKoch) [inline core/test/matrix/batch_csr.cpp:355]: same

File: core/test/matrix/batch_csr.cpp [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+    GKO_ASSERT_MTX_NEAR(csr_mats[1].get(), mat2.get(), 0.);
+}
+
+
+TYPED_TEST(Csr, CanBeListConstructed)
+{
+    using value_type = typename TestFixture::value_type;
+    using index_type = typename TestFixture::index_type;
+    using BatchCsrMtx = typename TestFixture::BatchCsrMtx;
+    using CsrMtx = typename TestFixture::CsrMtx;
```
Comment (MarcelKoch) [inline core/test/matrix/batch_csr.cpp:380]:
```suggestion
```

File: core/test/matrix/batch_csr.cpp [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+
+    this->assert_equal_to_original_sparse_mtx(m.get());
+}
+
+
+TYPED_TEST(Csr, ThrowsForDataWithDifferentNnz)
+{
+    using value_type = typename TestFixture::value_type;
+    using index_type = typename TestFixture::index_type;
+    using BatchCsrMtx = typename TestFixture::BatchCsrMtx;
```
Comment (MarcelKoch) [inline core/test/matrix/batch_csr.cpp:483]:
```suggestion
```

File: core/test/matrix/batch_csr.cpp [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+        gko::batch::detail::assert_same_sparsity_in_batched_data(vec_data),
+        gko::NotImplemented);
+}
+
+
+TYPED_TEST(Csr, ThrowsForDataWithDifferentSparsity)
+{
+    using value_type = typename TestFixture::value_type;
+    using index_type = typename TestFixture::index_type;
+    using BatchCsrMtx = typename TestFixture::BatchCsrMtx;
```
Comment (MarcelKoch) [inline core/test/matrix/batch_csr.cpp:505]:
```suggestion
```

File: reference/test/matrix/batch_csr_kernels.cpp [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+
+TYPED_TEST(Csr, AppliesToBatchMultiVector)
+{
+    using T = typename TestFixture::value_type;
+
+    this->mtx_0->apply(this->b_0.get(), this->x_0.get());
+
+    this->mtx_00->apply(this->b_00.get(), this->x_00.get());
+    this->mtx_01->apply(this->b_01.get(), this->x_01.get());
+    auto res = gko::batch::unbatch<gko::batch::MultiVector<T>>(this->x_0.get());
```
Comment (MarcelKoch) [inline reference/test/matrix/batch_csr_kernels.cpp:125]:
```suggestion

    auto res = gko::batch::unbatch<gko::batch::MultiVector<T>>(this->x_0.get());
```
for AAA

Comment (MarcelKoch) [reply reference/test/matrix/batch_csr_kernels.cpp:125]: More generally, AAA seems to not be followed correctly in this file.

**Review by MarcelKoch — CHANGES_REQUESTED** [review]
looks good, only some smaller comments

**Comment by pratikvn** [comment]
format!

File: core/matrix/csr.cpp [hunk 13]
```diff
@@ -443,8 +443,26 @@ void Csr<ValueType, IndexType>::read(device_mat_data&& data)
     auto arrays = data.empty_out();
     this->row_ptrs_.resize_and_reset(size[0] + 1);
     this->set_size(size);
-    this->values_ = std::move(arrays.values);
-    this->col_idxs_ = std::move(arrays.col_idxs);
+    // Need to copy instead of move, as move into view data is not supported.
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply core/matrix/csr.cpp:446]: No, move into view is not supported. It does not make sense to move data from an owning array to a non-owning one. I think the behaviour in array is correct. To have correct behaviour, we can either have all cases just copy, or copy only when the arrays of `this` are non-owning, and I think it is better to avoid deep copies wherever possible, so I think the implementation here is a good solution.

File: reference/test/matrix/batch_csr_kernels.cpp (hunk 12 above)
Comment (pratikvn) [reply reference/test/matrix/batch_csr_kernels.cpp:97]: I believe it should be. Act is only the batch matrix apply. The normal csr matrix apply comes in the assert block.

Comment (MarcelKoch) [reply reference/test/matrix/batch_csr_kernels.cpp:97]: Ok, I got a bit confused and thought the `mtx_0->apply` was the arrange part. But still I think we usually group applies of the to-test object together with the reference object. We do this in the common tests (reference and device apply) and in the MPI tests. So I would still suggest grouping all applies into the act category and then the unbatch into the assert caategory.

File: core/matrix/csr.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply core/matrix/csr.cpp:446]: I will create that PR.

[the remark this answers is not in the document]
Comment (pratikvn) [reply core/matrix/csr.cpp:446]: I think the behaviour of `const device_mat_data&` overload does copy. This function is just a helper that reduces code duplication by only taking in a temporary object. 

I agree that it is an edge-case, but it is very useful in implementing read for batched objects. If and when we implement non-uniform batch objects, I think we should follow a similar approach.

File: reference/test/matrix/batch_csr_kernels.cpp (hunk 12 above)
Comment (pratikvn) [reply reference/test/matrix/batch_csr_kernels.cpp:97]: I think that defeats the purpose of the AAA structure. IMO, the Act block should only contain function calls/declarations that are being tested, otherwise it is unclear what is being tested.

File: core/matrix/csr.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply core/matrix/csr.cpp:446]: Sadly the `const device_mat_data&` overload only copies the input, but it does not copy into the matrix's arrays, it moves again. So IMO, making the `const device_mat_data&` overload also copy the arrays would fix the issue here, while keeping the current move semantics.

[the remark this answers is not in the document]
Comment (MarcelKoch) [reply core/matrix/csr.cpp:446]: I would suggest that the input itself is not copied, but instead only the arrays are copied. This should have no overhead.

Moving data into views doesn't fail, it does exactly what is should do. The move overload takes ownership of the input data, and thus the matrix takes ownership of the stored arrays. So afterwards the matrix can't have array views anymore. I agree that we should clarify this, but I think this is the correct behavior.

[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/csr.cpp:446]: I tend to agree, reading into a view was never something we really considered before. This only matters for matrix types that reuse the arrays anyways (Csr and Coo), so it's not a notable maintenance overhead to duplicate a tiny bit of the implementation. We should have tests for both cases in place though

[the remark this answers is not in the document]
Comment (pratikvn) [reply core/matrix/csr.cpp:446]: I am fine with copying the arrays instead of copying the input. As we probably do that in another PR, maybe we push this PR forward with these changes ?

**Comment by pratikvn** [comment]
format!

File: dpcpp/matrix/batch_csr_kernels.hpp.inc [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    const auto val = mat.values;
+    const auto col = mat.col_idxs;
+    for (int row = item_ct1.get_local_linear_id(); row < num_rows;
+         row += item_ct1.get_local_range().size()) {
+        auto temp = zero<ValueType>();
+        for (auto nnz = mat.row_ptrs[row]; nnz < mat.row_ptrs[row + 1]; nnz++) {
+            const auto col_idx = col[nnz];
+            temp += alpha * val[nnz] * b[col_idx];
+        }
+        x[row] = temp + beta * x[row];
```
Comment (yhmtsai) [inline dpcpp/matrix/batch_csr_kernels.hpp.inc:42]:
```suggestion
            temp += val[nnz] * b[col_idx];
        }
        x[row] = alpha * temp + beta * x[row];
```

File: reference/test/solver/batch_bicgstab_kernels.cpp [hunk 15]
```diff
@@ -247,6 +249,42 @@ TYPED_TEST(BatchBicgstab, CanSolveEllSystem)
 }
 
 
+TYPED_TEST(BatchBicgstab, CanSolveCsrSystem)
```
Comment (yhmtsai) [inline reference/test/solver/batch_bicgstab_kernels.cpp:252]: Maybe this can be change to the uniform test for different matrix format.
(not need in the pr)

File: include/ginkgo/core/matrix/batch_csr.hpp [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+    {
+        return values_.get_size();
+    }
+
+    /**
+     * Returns the number of stored elements in each batch item.
+     *
+     * @return the number of stored elements per batch item.
+     */
+    size_type get_num_elements_per_item() const noexcept
```
Comment (yhmtsai) [inline include/ginkgo/core/matrix/batch_csr.hpp:167]: we use `get_num_stored_elements()` in other place.
should it be `get_num_stored_elements_per_item()` for consistence? it's quite long though

File: include/ginkgo/core/matrix/batch_csr.hpp [hunk 17]
```diff
@@ -0,0 +1,341 @@
...
+namespace gko {
+namespace batch {
+namespace matrix {
+
+
+/**
+ * Csr is a general sparse matrix format that stores the column indices for each
+ * nonzero entry and a cumulative sum of the number of nonzeros in each row. It
+ * is one of the most popular sparse matrix formats due to its versatility and
+ * ability to store a wide range of sparsity patterns in an efficient fashion.
```
Comment (yhmtsai) [inline include/ginkgo/core/matrix/batch_csr.hpp:33]: not update documentation with batch like batchEll?

File: include/ginkgo/core/matrix/batch_csr.hpp [hunk 18]
```diff
@@ -0,0 +1,341 @@
...
+        : EnableBatchLinOp<Csr>(exec, size),
+          values_{exec, std::forward<ValuesArray>(values)},
+          col_idxs_{exec, std::forward<ColIdxsArray>(col_idxs)},
+          row_ptrs_{exec, std::forward<RowPtrsArray>(row_ptrs)}
+    {
+        // Ensure that the value and col_idxs arrays have the correct size
+        auto max_num_elems = this->get_common_size()[0] *
+                             this->get_common_size()[1] *
+                             this->get_num_batch_items();
+        GKO_ASSERT(values_.get_size() <= max_num_elems);
```
Comment (yhmtsai) [inline include/ginkgo/core/matrix/batch_csr.hpp:317]: do you need to check this? it is just unused from the batch csr

Comment (yhmtsai) [reply include/ginkgo/core/matrix/batch_csr.hpp:317]: ah, the get_num_stored_elements is based on values.size(), so it is required to only store the batch value

File: reference/test/matrix/batch_csr_kernels.cpp [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    Csr()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx_0(gko::batch::initialize<BMtx>(
+              {{I<T>({1.0, -1.0, 1.5}), I<T>({-2.0, 2.0, 3.0})},
+               {{1.0, -2.0, -0.5}, {1.0, -2.5, 4.0}}},
+              exec, 6)),
+          mtx_00(gko::initialize<CsrMtx>(
+              {I<T>({1.0, -1.0, 1.5}), I<T>({-2.0, 2.0, 3.0})}, exec)),
+          mtx_01(gko::initialize<CsrMtx>(
+              {I<T>({1.0, -2.0, -0.5}), I<T>({1.0, -2.5, 4.0})}, exec)),
```
Comment (yhmtsai) [inline reference/test/matrix/batch_csr_kernels.cpp:46]: maybe use sparse data not the full dense one?

File: reference/test/matrix/batch_csr_kernels.cpp [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    std::unique_ptr<DenseMtx> b_01;
+    std::unique_ptr<BMVec> x_0;
+    std::unique_ptr<DenseMtx> x_00;
+    std::unique_ptr<DenseMtx> x_01;
+
+    std::ranlux48 rand_engine;
+};
+
+
+TYPED_TEST_SUITE(Csr, gko::test::ValueTypes, TypenameNameGenerator);
```
Comment (yhmtsai) [inline reference/test/matrix/batch_csr_kernels.cpp:86]:
```suggestion
TYPED_TEST_SUITE(Csr, gko::test::ValueTypes, TypenameNameGenerator);
```

File: reference/test/solver/batch_bicgstab_kernels.cpp (hunk 15 above)
Comment (pratikvn) [reply reference/test/solver/batch_bicgstab_kernels.cpp:252]: Yes, maybe I will handle it in another PR

File: include/ginkgo/core/matrix/batch_csr.hpp (hunk 16 above)
Comment (yhmtsai) [reply include/ginkgo/core/matrix/batch_csr.hpp:167]: I think we try to avoid nnz because the stored element can be zeros

File: include/ginkgo/core/matrix/batch_csr.hpp (hunk 17 above)
Comment (yhmtsai) [reply include/ginkgo/core/matrix/batch_csr.hpp:33]: It's like CSR not batchCSR

Comment (pratikvn) [reply include/ginkgo/core/matrix/batch_csr.hpp:33]: The note below clarifies that this is a batched matrix fomat, and how the batches are stored. I think being within the namespace also makes that clear. It is similar to `batch::Ell`. But I think can add more clarification to make it clear.

File: core/test/matrix/batch_csr.cpp [hunk 21]
```diff
@@ -0,0 +1,511 @@
...
+        {2, 3}, {{0, 0, 1.0}, {1, 1, 2.0}, {1, 2, 3.0}}));
+
+    auto m = gko::batch::read<value_type, index_type, BatchCsrMtx>(this->exec,
+                                                                   vec_data, 3);
+
+    this->assert_equal_to_original_sparse_mtx(m.get());
+}
+
+
+TYPED_TEST(Csr, ThrowsForDataWithDifferentNnz)
```
Comment (MarcelKoch) [inline core/test/matrix/batch_csr.cpp:450]: This and the next test are already done in batch_ell. And generally, these tests are not matrix type specific, so they should be in their own file.

File: reference/test/matrix/batch_csr_kernels.cpp [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+    this->mtx_01->apply(this->b_01.get(), this->x_01.get());
+    auto res = gko::batch::unbatch<gko::batch::MultiVector<T>>(this->x_0.get());
+    GKO_ASSERT_MTX_NEAR(res[0].get(), this->x_00.get(), r<T>::value);
+    GKO_ASSERT_MTX_NEAR(res[1].get(), this->x_01.get(), r<T>::value);
+}
+
+
+TYPED_TEST(Csr, AppliesLinearCombinationToBatchMultiVector)
+{
+    using BMtx = typename TestFixture::BMtx;
```
Comment (MarcelKoch) [inline reference/test/matrix/batch_csr_kernels.cpp:119]: unused
```suggestion
```

**Review by MarcelKoch — APPROVED** [review]
LGTM. I think the testing strategy should be revisited at some point. A lot of tests in core test functionality that is composed of smaller parts, but the implementation is the same for all different matrix types. For example, `batch::write` only requires that the type has `create_const_view_for_item`. Because of that many tests are repetetive and could probably be unified.

**Comment by pratikvn** [comment]
As the previous pipeline had fully passed, and I only removed a couple of tests, I will just go ahead and merge this PR, not waiting for the whole pipeline to complete.
